restic 0.13.0 tests fail on 32-bit architectures

While upgrading the Alpine Linux restic package from 0.12.0 to 0.13.0 I noticed a failure of the following test case on 32-bit architectures (armv7 and x86):

https://github.com/restic/restic/blob/7080fed7ae0eb660e0ba9f352ccf3765ea2e9ed8/internal/fs/setflags_linux_test.go#L57-L67

This test case was added in commit 7080fed7ae0eb660e0ba9f352ccf3765ea2e9ed8 (CC: @greatroar).

The output for the failed test looks as follows:

```
go test ./cmd/... ./internal/...
go: downloading github.com/google/go-cmp v0.5.6
# github.com/restic/restic/internal/fs [github.com/restic/restic/internal/fs.test]
internal/fs/setflags_linux_test.go:62:24: unix.BTRFS_SUPER_MAGIC (untyped int constant 2435016766) overflows int32
ok  	github.com/restic/restic/cmd/restic	18.070s
ok  	github.com/restic/restic/internal/archiver	7.425s
ok  	github.com/restic/restic/internal/backend	3.505s
ok  	github.com/restic/restic/internal/backend/azure	0.006s
ok  	github.com/restic/restic/internal/backend/b2	0.008s
ok  	github.com/restic/restic/internal/backend/dryrun	0.005s
ok  	github.com/restic/restic/internal/backend/gs	0.008s
ok  	github.com/restic/restic/internal/backend/local	2.878s
ok  	github.com/restic/restic/internal/backend/location	0.012s
ok  	github.com/restic/restic/internal/backend/mem	2.560s
ok  	github.com/restic/restic/internal/backend/rclone	0.015s
ok  	github.com/restic/restic/internal/backend/rest	0.013s
ok  	github.com/restic/restic/internal/backend/s3	0.006s
ok  	github.com/restic/restic/internal/backend/sftp	0.007s
ok  	github.com/restic/restic/internal/backend/swift	0.007s
ok  	github.com/restic/restic/internal/backend/test	3.021s
ok  	github.com/restic/restic/internal/bloblru	0.006s
ok  	github.com/restic/restic/internal/cache	1.122s
ok  	github.com/restic/restic/internal/checker	2.603s
ok  	github.com/restic/restic/internal/crypto	0.619s
ok  	github.com/restic/restic/internal/debug	0.006s [no tests to run]
ok  	github.com/restic/restic/internal/dump	0.165s
?   	github.com/restic/restic/internal/errors	[no test files]
ok  	github.com/restic/restic/internal/filter	0.510s
FAIL	github.com/restic/restic/internal/fs [build failed]
ok  	github.com/restic/restic/internal/fuse	0.059s
ok  	github.com/restic/restic/internal/hashing	0.187s
ok  	github.com/restic/restic/internal/limiter	0.006s
?   	github.com/restic/restic/internal/migrations	[no test files]
?   	github.com/restic/restic/internal/mock	[no test files]
ok  	github.com/restic/restic/internal/options	0.025s
ok  	github.com/restic/restic/internal/pack	0.018s
ok  	github.com/restic/restic/internal/repository	6.739s
ok  	github.com/restic/restic/internal/restic	1.944s
ok  	github.com/restic/restic/internal/restorer	0.069s
?   	github.com/restic/restic/internal/selfupdate	[no test files]
?   	github.com/restic/restic/internal/test	[no test files]
ok  	github.com/restic/restic/internal/textfile	0.017s
ok  	github.com/restic/restic/internal/ui	0.014s
?   	github.com/restic/restic/internal/ui/backup	[no test files]
ok  	github.com/restic/restic/internal/ui/progress	0.143s
?   	github.com/restic/restic/internal/ui/signals	[no test files]
ok  	github.com/restic/restic/internal/ui/table	0.027s
ok  	github.com/restic/restic/internal/ui/termstatus	0.006s
ok  	github.com/restic/restic/internal/walker	0.010s
FAIL
make: *** [Makefile:12: test] Error 2
```

A change along the lines of:

```diff
diff -upr restic-0.13.0.orig/internal/fs/setflags_linux_test.go restic-0.13.0/internal/fs/setflags_linux_test.go
--- restic-0.13.0.orig/internal/fs/setflags_linux_test.go	2022-03-29 21:31:21.121497170 +0200
+++ restic-0.13.0/internal/fs/setflags_linux_test.go	2022-03-29 21:31:35.918229947 +0200
@@ -59,7 +59,7 @@ func supportsNoatime(t *testing.T, f *os
 	err := unix.Fstatfs(int(f.Fd()), &fsinfo)
 	rtest.OK(t, err)
 
-	return fsinfo.Type == unix.BTRFS_SUPER_MAGIC ||
+	return int64(fsinfo.Type) == int64(unix.BTRFS_SUPER_MAGIC) ||
 		fsinfo.Type == unix.EXT2_SUPER_MAGIC ||
 		fsinfo.Type == unix.EXT3_SUPER_MAGIC ||
 		fsinfo.Type == unix.EXT4_SUPER_MAGIC ||
```

fixes the test failure, but not sure if this is the correct way to address this issue, hence reporting it as a bug.

--- a/internal/fs/setflags_linux_test.go
+++ b/internal/fs/setflags_linux_test.go
@@ -59,9 +59,13 @@ func supportsNoatime(t *testing.T, f *os.File) bool {
 	err := unix.Fstatfs(int(f.Fd()), &fsinfo)
 	rtest.OK(t, err)
 
-	return fsinfo.Type == unix.BTRFS_SUPER_MAGIC ||
-		fsinfo.Type == unix.EXT2_SUPER_MAGIC ||
-		fsinfo.Type == unix.EXT3_SUPER_MAGIC ||
-		fsinfo.Type == unix.EXT4_SUPER_MAGIC ||
-		fsinfo.Type == unix.TMPFS_MAGIC
+	// The funky cast works around a compiler error on 32-bit archs:
+	// "unix.BTRFS_SUPER_MAGIC (untyped int constant 2435016766) overflows int32".
+	// https://github.com/golang/go/issues/52061
+	typ := int64(uint(fsinfo.Type))
+	return typ == unix.BTRFS_SUPER_MAGIC ||
+		typ == unix.EXT2_SUPER_MAGIC ||
+		typ == unix.EXT3_SUPER_MAGIC ||
+		typ == unix.EXT4_SUPER_MAGIC ||
+		typ == unix.TMPFS_MAGIC
 }

